### PR TITLE
[tracing] fix ScopedTraceBlock reordering bug

### PR DESCRIPTION
**Pull request by nickgg**
*Description*: I had noticed that some of the Scoped TraceEvents had incorrect timestamps, often with zero duration. There were two causes of this:
* In one case I forgot to name the object, which means it's destroyed immediately.
* Since the trace block object is never referenced the compiler was reordering the end timestamp to immediately after the creation of the object (rather than when the scope ended). Adding a memory fence prevents this reordering.
*Testing*: unit tests & a manual test inspecting trace events.
*Documentation*:

### CODE REVIEW

File: lib/Backends/TraceEvents.cpp [hunk 1]
```diff
@@ -121,6 +121,8 @@ ScopedTraceBlock::ScopedTraceBlock(TraceContext *context, llvm::StringRef name)
   if (context_) {
     context_->logTraceEvent(name_, TraceEvent::BeginType, std::move(args_));
   }
+
```
Comment (nadavrot) [inline lib/Backends/TraceEvents.cpp]: I think that we need a comment here that explains what's going on and why do we have this fence.

File: lib/Backends/TraceEvents.cpp [hunk 2]
```diff
@@ -132,6 +134,7 @@ ScopedTraceBlock &ScopedTraceBlock::addArg(llvm::StringRef key,
 }
 
 void ScopedTraceBlock::end() {
+  std::atomic_signal_fence(std::memory_order_seq_cst);
```
Comment (nadavrot) [inline lib/Backends/TraceEvents.cpp]: Also here

**Comment by rdzhabarov** [comment]
> In one case I forgot to name the object, which means it's destroyed immediately.
was it merged in some other PR? (do not see it here currently)

**Comment by nadavrot** [comment]
Perfect!

**Comment by nickgg** [comment]
> was it merged in some other PR? (do not see it here currently)

@rdzhabarov yes, good catch - it was replaced by the TRACE_EVENT_SCOPE macro in #2854.

**Pull request merged**